### DISCUSSION TITLE
migration of Das2 (Java) library to GitHub from saturn

### DISCUSSION THREAD

**Discussion by jbfaden** [das-developers/discussions#5] [General]
We're planning to shut down the SVN server "saturn" at U. Iowa and move projects over to this part of GitHub.  This will affect a number of projects, including Autoplot and APL's MIDL software.  Both Autoplot and MIDL are on SVN servers and use SVN externals to bring in the code.  I (Jeremy) plan to move Autoplot to GitHub in the next year, and I don't know what APL's plans are.  It may be that they are simply copying versions of Das2 into their library without using an externals mechanism.

Das2 is stored in four Netbeans projects:
1. dasCoreUtil - utilities like FileSystems and ProgressMonitors
2. dasCoreDatum - Units, Datums, DatumRanges, and TimeUtil
3. QDataSet - data model from PaPCo which was brought in around 2011 and has grown along with Autoplot.  Depends on 1 and 2.
4. dasCore - graphics and events, file storage model, old das2stream library.  Depends on 1, 2, and 3.

The goals are:
1. to try and preserve the revision history with version numbers as much as possible. 
2. move QDataSet from Autoplot's source into das2, as it is used extensively.

Making things a little trickier is the fork of the das2 library which happened around 2011, which replaced the old Das2 data model with QDataSet.  This has been the active fork for the past 10 years, and the few changes made the old fork have been moved to the new fork.  However, there may still be applications which use the old version of dasCore.  Chris already has demonstrated that these could be made into two branches on GitHub.

Also to consider for the das-developers home:
1. ScienceDataInterfaces, which is a more abstract API for modeling data
2. QDataSetScienceDataInterface, which implements ScienceDataInterfaces using QDataSet.
3. QStream was intended to replace das2streams and might be moved as well, but it has never had much use in the Radio and Plasma Wave Group and could be left as part of Autoplot.
4. https://github.com/hapi-server/uri-templates is a TimeParser-like library extracted which works independently of dasCoreDatum, and I've intended to refactor dasCoreUtil and dasCore to use it.

**Comment by cpiker** [comment]
Hi @jbfaden, so I'll rework the test repository with the following changes:

* Move the `autoplot` branch name to `main`
* Fold in the QStream netbeans project.

We do output qstreams over at U. Iowa, though since [das2C](https://github.com/das-developers/das2C) can't parse qstreams they aren't used as much.  Since QStream logically fits in next to QDataSet, I'd prefer to include it too if that's okay.  I looked around a bit but couldn't find ScienceDataInterfaces or QDataSetScienceDataInterface.  If you want those included as well just let me know where they live, maybe @jpeachey knows where to find them?

One question for everyone.  It's customary to put a license file in the root of the repository.  Looking over the source code I've seen different license statements, and many files have no license mentioned at all.  As long as the original programmers are contactable it would probably be best to settle on one license for the whole repository.  The main reason I created the initial test repository as a private one to start with is I'm not sure which license should be listed.  Since I haven't contributed much code to the Java libraries I'll just go along with whatever license the majority contributors prefer.

**Reply by cpiker** [reply]
I've recreated the repository to add in many of your suggested changes.  The most actively maintained and improved branch is named `main`, and QStream is included.

The das2 apps branch is named `original` and I've cleaned up non-mavenized source artifacts in the original branch since it, and the marsis app, use maven as the build tool.

Some work is still needed to get the license straight.  I didn't notice the class-path exception in the license file, so that probably should get added.

**Reply by jbfaden** [reply]
The Autoplot project has a license statement in its root, which I think is all that is needed.  You might be right about the need for a classpath exception, maybe this version works: https://openjdk.java.net/legal/gplv2+ce.html .

**Comment by cpiker** [comment]
That license looks a bit stricter than the one in the root of Autoplot.  The Autoplot one is LGPL, but the one above is GPL based so it might scare folks away.   If you do decide to use the one above, we'd of course have to strip out the Oracle reference and replace it with something else.  I assume the primary concern is not to accidentally conflict with MIDL or Autoplot.

For my code I've been using LGPL in core libs and MIT in higher level stuff (examples, app-code).  If it's any help, the SPDF asked me to use Apache 2.0 for my D prototypes to CDF.  They've probably considered license implications much more than I have.

**Comment by jbfaden** [comment]
The intent was to have a GPL-like license that was able to connect to non-GPL things, so LGPL or the GPL plus classpath exception would probably work.

**Comment by cpiker** [comment]
@jbfaden I'm working to build a copy of Autoplot based off the das2java repository and have noticed a few  libraries in APlibs that are needed in dasCore and dasCoreUtil.  I'd like to copy these over.  There seems to be two basic layouts for doing this.  One is to have a specific top-level library folder and put everything in there such as:

```
   dasCore
   dasCoreDatum
   dasCoreUtil
   QDataSet
   QStream
   lib     <--- put all upstream jars here
```
or the per-project version:
```
   dasCore
    |- lib      <--- put some upstream jars here
   dasCoreDatum
    |- lib     <--- and some here
   dasCoreUtil
    |- lib     <--- and some here
   QDataSet
    |- lib     <--- and some here
   QStream
    |- lib     <--- and some here

```
Which would you prefer?  The first one seems easier to manage, but might not follow other norms.  Also If using a top level lib collection, should that folder be a netbeans project as well?

**Reply by jbfaden** [reply]
I think the second is better, since it's more conventional.  (Netbeans conventions, that is...)

**Reply by cpiker** [reply]
Sounds good, that's the arrangement I'll test now.
